### ISSUE TITLE
F0816 13:49:00.962771  8084 layer_factory.hpp:81] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: NdConvolution 

### ISSUE THREAD

**Issue by lishanlu136** [chuckcho/video-caffe#36]
When I train the example network based video frames. An error occurred as follows:

``````
F0816 13:49:00.962771  8084 layer_factory.hpp:81] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: NdConvolution (known types: AbsVal, Accuracy, ArgMax, BNLL, BatchNorm, BatchReindex, Bias, Concat, ContrastiveLoss, Convolution, Crop, Data, Deconvolution, Dropout, DummyData, ELU, Eltwise, Embed, EuclideanLoss, Exp, Filter, Flatten, HDF5Data, HDF5Output, HingeLoss, Im2col, ImageData, InfogainLoss, InnerProduct, Input, LRN, LSTM, LSTMUnit, Log, MVN, MemoryData, MultinomialLogisticLoss, PReLU, Parameter, Pooling, Power, RNN, ReLU, Reduction, Reshape, SPP, Scale, Sigmoid, SigmoidCrossEntropyLoss, Silence, Slice, Softmax, SoftmaxWithLoss, Split, TanH, Threshold, Tile, VideoData, WindowData)
*** Check failure stack trace: ***
    @     0x7ff2be269daa  (unknown)
    @     0x7ff2be269ce4  (unknown)
    @     0x7ff2be2696e6  (unknown)
    @     0x7ff2be26c687  (unknown)
    @     0x7ff2be8d9dc4  caffe::Net<>::Init()
    @     0x7ff2be8db7e5  caffe::Net<>::Net()
    @     0x7ff2be8becdf  caffe::Solver<>::InitTrainNet()
    @     0x7ff2be8bf1fb  caffe::Solver<>::Init()
    @     0x7ff2be8bf668  caffe::Solver<>::Solver()
    @     0x7ff2be8edd93  caffe::Creator_SGDSolver<>()
    @           0x410fe1  caffe::SolverRegistry<>::CreateSolver()
    @           0x40a613  train()
    @           0x406d22  main
    @     0x7ff2bd577f45  (unknown)
    @           0x40726c  (unknown)
    @              (nil)  (unknown)```
I did not install cudnn, because the GPU version is too low. But I guess the error not because of I did not install cudnn. How can I do to solve this error？
``````

**Comment by chuckcho** [comment]
Yes -- it's likely related to CUDNN not installed. `NdConvolution` needs CUDNN

**Comment by lishanlu136** [comment]
Thank you for your reply. I look at the caffe.proto, and "NdConvolution" is not defined in layer type. When I replaced NdConvolution with Convolution, it works. But convolution_param and Pooling_Parameter were not work, it only works when Pad, kernel size, and stride are all given as a single value for equal dimensions in all spatial dimensions. Could you give me a new train example?

**Comment by chuckcho** [comment]
Sorry -- CUDA/CUDNN is required.

**Comment by codingsmallfish** [comment]
@chuckcho Hi, I have checked the installation of CUDA 7.0/CUDNN on my computer. However, it still shows `F0919 12:06:25.276957 29350 layer_factory.hpp:81] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: NdConvolution`. 

Can you provide some suggestion?

Thanks.

**Comment by chuckcho** [comment]
my feeling is that `cudnn` is somehow not enabled. can you run `make test.testbin && cd .. && build/test/test.testbin --gtest_filter=*CuDNN*:*Cudnn*`, for example, to see if cudnn tests are being done.

**Comment by codingsmallfish** [comment]
@chuckcho I didn't find `test.testbin` in `build/test` and when I run `make test.testbin` in `video-caffe` file, it returns `No rule to make target - test.testbin`. Actually the `build/test` file contains no file at all. Is that a problem? 
However I successfully run `make test`, and there's no error showing.

**Comment by chuckcho** [comment]
I was referring to "cmake" building. If you were using `Makefile`, do `make runtest` and check if "cudnn" related tests are done. e.g. `CudnnNdPoolingLayerTest` and `CuDNNConvolutionLayerTest`.

**Comment by codingsmallfish** [comment]
OK, thank you. I did `make runtest`, the program stopped with
`E0919 13:51:33.680763 12739 io.cpp:181] Cannot open a video file=src/caffe/test/test_data/UCF-101_Rowing_g16_c03.avi'
'F0919 13:51:33.681011 12739 video_data_layer.cpp:77] Check failed: read_video_result Could not load src/caffe/test/test_data/UCF-101_Rowing_g16_c03.avi at frame 0.`

I have checked the `.avi` file's existance and didn't see `CudnnNdPoolingLayerTest` or `CuDNNConvolutionLayerTest` yet.

**Comment by codingsmallfish** [comment]
Is there any probability this issue is related to opencv?

**Comment by chuckcho** [comment]
If you didn't see any CuDNN-related tests and your caffe was built without CuDNN support.

**Comment by chuckcho** [comment]
cudnn-related issue. closing.

**Comment by Hzzone** [comment]
@chuckcho I have compile `NdConvolution` on master branch successfully, however failed on refactor branch? I want to use hdf5 format file to train my data. Thank you for you help!
